File: packages/parent-domain/src/browser-policy-questionnaire-forest.ts
import { browserControlFullCatalogSettings } from './browser-control-full-catalog';
import {
  BrowserPolicyQuestionIds,
  type BrowserPolicyAnswerMap,
  type BrowserPolicyComputedFlagId,
  type BrowserPolicyCondition,
  type BrowserPolicyQuestion,
  type BrowserPolicyQuestionId,
  type BrowserPolicyQuestionState,
  type BrowserPolicySurface,
} from './browser-policy-questionnaire-forest-contract';
import {
  BrowserPolicyQuestionnaireCompactOrder,
  BrowserPolicyQuestions,
} from './browser-policy-questionnaire-forest-data';

export {
  BrowserPolicyDefaultAnswers,
  BrowserPolicyQuestionIds,
  type BrowserPolicyAnswerMap,
  type BrowserPolicyComputedFlagId,
  type BrowserPolicyCondition,
  type BrowserPolicyOption,
  type BrowserPolicyQuestion,
  type BrowserPolicyQuestionId,
  type BrowserPolicyQuestionState,
  type BrowserPolicySelectionMode,
  type BrowserPolicySurface,
} from './browser-policy-questionnaire-forest-contract';
export {
  BrowserPolicyQuestionnaireCompactOrder,
  BrowserPolicyQuestions,
} from './browser-policy-questionnaire-forest-data';

export const BrowserPolicyQuestionnaireForest = {
  questions: BrowserPolicyQuestions,
  compactOrder: BrowserPolicyQuestionnaireCompactOrder,
} as const;

export function browserPolicyQuestionById(id: BrowserPolicyQuestionId): BrowserPolicyQuestion {
  const question = BrowserPolicyQuestions.find((candidate) => candidate.id === id);
  if (question === undefined) {
    throw new Error(`Unknown browser policy question ${id}`);
  }
  return question;
}

export function browserPolicyQuestionStates(
  answers: BrowserPolicyAnswerMap,
  questions: readonly BrowserPolicyQuestion[] = BrowserPolicyQuestions
): BrowserPolicyQuestionState[] {
  return questions.map((question) => browserPolicyQuestionState(question, answers));
}

export function browserPolicyVisibleQuestions(
  answers: BrowserPolicyAnswerMap,
  surface: BrowserPolicySurface = 'rules'
): BrowserPolicyQuestion[] {
  const questionById = new Map(BrowserPolicyQuestions.map((question) => [question.id, question]));
  const orderedIds =
    surface === 'ai'
      ? BrowserPolicyQuestionIds.filter((id) => id.startsWith('A'))
      : BrowserPolicyQuestionnaireCompactOrder;
  return orderedIds
    .map((id) => questionById.get(id))
    .filter((question): question is BrowserPolicyQuestion => question !== undefined)
    .filter((question) => question.surface === surface)
    .filter((question) => browserPolicyQuestionState(question, answers).visible);
}

export function browserPolicyQuestionState(
  question: BrowserPolicyQuestion,
  answers: BrowserPolicyAnswerMap
): BrowserPolicyQuestionState {
  const neverShow = conditionListMatches(question.neverShowWhen, answers);
  const show = (question.showWhen ?? []).length === 0 || conditionsMatch(question.showWhen ?? [], answers);
  const visible = show && !neverShow;
  return {
    question,
    visible,
    disabled: visible && conditionListMatches(question.disabledWhen, answers),
    readonly: visible && conditionListMatches(question.readonlyWhen, answers),
  };
}

export function browserPolicyConditionMatches(
  condition: BrowserPolicyCondition,
  answers: BrowserPolicyAnswerMap
): boolean {
  const selected = (questionId: BrowserPolicyQuestionId) => answers[questionId] ?? [];
  switch (condition.kind) {
    case 'answer-equals':
      return selected(condition.questionId)[0] === condition.optionId;
    case 'answer-includes':
      return selected(condition.questionId).includes(condition.optionId);
    case 'answer-includes-any':
      return condition.optionIds.some((optionId) => selected(condition.questionId).includes(optionId));
    case 'answer-has-any-selected':
      return selected(condition.questionId).length > 0;
    case 'computed-flag':
      return browserPolicyComputedFlag(condition.flagId, answers);
    case 'all':
      return conditionsMatch(condition.conditions, answers);
    case 'any':
      return condition.conditions.some((candidate) => browserPolicyConditionMatches(candidate, answers));
    case 'not':
      return !browserPolicyConditionMatches(condition.condition, answers);
  }
}

export function browserPolicyComputedFlag(
  flagId: BrowserPolicyComputedFlagId,
  answers: BrowserPolicyAnswerMap
): boolean {
  return BrowserPolicyComputedFlagEvaluators[flagId](answers);
}

function browserPolicyRootAnswer(answers: BrowserPolicyAnswerMap): string {
  return answers['1.1']?.[0] ?? 'off';
}

function browserPolicyHas(
  answers: BrowserPolicyAnswerMap,
  questionId: BrowserPolicyQuestionId,
  optionId: string
): boolean {
  return (answers[questionId] ?? []).includes(optionId);
}

function browserPolicyHasAny(
  answers: BrowserPolicyAnswerMap,
  questionId: BrowserPolicyQuestionId,
  optionIds: readonly string[]
): boolean {
  return optionIds.some((optionId) => browserPolicyHas(answers, questionId, optionId));
}

function browserPolicyCount(answers: BrowserPolicyAnswerMap, questionId: BrowserPolicyQuestionId): number {
  return answers[questionId]?.length ?? 0;
}

function browserPolicyEmergencyOverrideActive(answers: BrowserPolicyAnswerMap): boolean {
  return browserPolicyHasAnyRoot(answers, ['emergency-allow', 'emergency-block']);
}

function browserPolicyHasAnyRoot(answers: BrowserPolicyAnswerMap, rootValues: readonly string[]): boolean {
  return rootValues.includes(browserPolicyRootAnswer(answers));
}

function browserPolicyAskParentExists(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAnyAskParentTrigger(answers) ||
    browserPolicyHasAny(answers, '10.2', ['allow-once', 'custom-window']) ||
    browserPolicyHas(answers, '11.2', 'extension') ||
    browserPolicyRootAnswer(answers) === 'emergency-block'
  );
}

function browserPolicyHasAnyAskParentTrigger(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAny(answers, '1.2', ['ask-parent']) ||
    browserPolicyHasAny(answers, '2.3', ['ask-parent']) ||
    browserPolicyHasAny(answers, '4.1', ['ask-parent']) ||
    browserPolicyHasAny(answers, '5.2', ['ask-parent']) ||
    browserPolicyHasAny(answers, '5.3', ['ask-parent']) ||
    browserPolicyHasAny(answers, '6.1', ['ask-parent']) ||
    browserPolicyHasAny(answers, '7.1', ['ask-parent']) ||
    browserPolicyHasAny(answers, '8.1', ['ask-parent']) ||
    browserPolicyHasAny(answers, '9.1', ['ask-parent']) ||
    browserPolicyHasAny(answers, '9.3', ['ask-parent']) ||
    browserPolicyHasAny(answers, '15.2', ['ask-parent'])
  );
}

function browserPolicyExactEvidenceSelected(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAny(answers, '5.1', ['exact-url', 'search-terms']) ||
    browserPolicyHasAny(answers, '13.1', ['exact-url', 'search-term']) ||
    browserPolicyHas(answers, '3.1', 'managed-exact')
  );
}

function browserPolicyManagedBrowserRequired(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAny(answers, '3.1', ['managed-exact', 'managed-all']) ||
    browserPolicyHas(answers, '6.1', 'require-managed')
  );
}

function browserPolicySetupRelevant(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAny(answers, '3.1', ['prefer-managed', 'managed-exact', 'managed-all']) ||
    browserPolicyHasAny(answers, '2.2', ['standard', 'strict', 'custom']) ||
    browserPolicyHasAny(answers, '2.3', ['ask-parent', 'block-until-approved']) ||
    browserPolicyComputedFlag('unsupportedCapabilityRelevant', answers)
  );
}

function browserPolicyEvidencePrivacyVisible(answers: BrowserPolicyAnswerMap): boolean {
  return browserPolicyRootAnswer(answers) === 'on' && browserPolicyHasAnyEvidenceTrigger(answers);
}

function browserPolicyHasAnyEvidenceTrigger(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyCount(answers, '2.1') > 0 ||
    browserPolicyCount(answers, '5.1') > 0 ||
    browserPolicyCount(answers, '7.1') > 0 ||
    browserPolicyCount(answers, '8.1') > 0 ||
    browserPolicyCount(answers, '9.1') > 0 ||
    browserPolicyComputedFlag('reportsEnabled', answers)
  );
}

function browserPolicyNotificationEventsRelevant(answers: BrowserPolicyAnswerMap): boolean {
  return browserPolicyRootAnswer(answers) !== 'off' && browserPolicyHasAnyNotificationTrigger(answers);
}

function browserPolicyHasAnyNotificationTrigger(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAny(answers, '2.3', ['notify-parent', 'ask-parent', 'block-until-approved']) ||
    browserPolicyHasAny(answers, '4.1', [
      'warn',
      'notify-parent',
      'ask-parent',
      'close',
      'close-open-managed',
      'block-launch',
    ]) ||
    browserPolicyHasAny(answers, '6.1', ['warn', 'ask-parent', 'block', 'close-browser']) ||
    browserPolicyHasAny(answers, '9.1', ['notify-parent', 'ask-parent', 'block-risky', 'block-all-approved']) ||
    browserPolicyComputedFlag('limitExists', answers) ||
    browserPolicyHasAnyRoot(answers, ['paused', 'emergency-allow', 'emergency-block'])
  );
}

function browserPolicyUnsupportedCapabilityRelevant(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyHasAny(answers, '2.1', ['strict', 'custom']) ||
    browserPolicyComputedFlag('managedBrowserRequired', answers) ||
    browserPolicyHasAny(answers, '5.1', ['exact-url', 'search-terms', 'downloads', 'video']) ||
    browserPolicyHasAny(answers, '6.1', ['block', 'close-browser', 'require-managed'])
  );
}

function browserPolicyStoredBrowserDataExists(answers: BrowserPolicyAnswerMap): boolean {
  return (
    browserPolicyComputedFlag('reportsEnabled', answers) ||
    browserPolicyComputedFlag('auditEnabled', answers) ||
    browserPolicyCount(answers, '13.1') > 0 ||
    browserPolicyCount(answers, '12.1') > 0
  );
}

const BrowserPolicyComputedFlagEvaluators: Record<
  BrowserPolicyComputedFlagId,
  (answers: BrowserPolicyAnswerMap) => boolean
> = {
  policyIsOff: (answers) => browserPolicyRootAnswer(answers) === 'off',
  policyIsOn: (answers) => browserPolicyRootAnswer(answers) === 'on',
  policyPaused: (answers) => browserPolicyRootAnswer(answers) === 'paused',
  emergencyOverrideActive: browserPolicyEmergencyOverrideActive,
  askParentExists: browserPolicyAskParentExists,
  limitExists: (answers) =>
    browserPolicyHas(answers, '1.2', 'limit') ||
    browserPolicyHas(answers, '6.1', 'limit-time') ||
    browserPolicyHas(answers, '8.1', 'limit'),
  downloadsSelected: (answers) => browserPolicyHas(answers, '5.1', 'downloads'),
  searchSelected: (answers) => browserPolicyHasAny(answers, '5.1', ['search-terms', 'safe-search']),
  videoSelected: (answers) => browserPolicyHas(answers, '5.1', 'video'),
  exactEvidenceSelected: browserPolicyExactEvidenceSelected,
  managedBrowserRequired: browserPolicyManagedBrowserRequired,
  reportsEnabled: (answers) => (answers['14.1'] ?? []).some((optionId) => optionId !== 'policy-status'),
  auditEnabled: (answers) => browserPolicyHasAny(answers, '18.1', ['minimal', 'standard', 'detailed', 'custom']),
  setupRelevant: browserPolicySetupRelevant,
  classificationServiceReferenced: (answers) =>
    browserPolicyHas(answers, '5.1', 'category') || browserPolicyHas(answers, '5.2', 'classification-service'),
  multiTargetActionMatrixRelevant: (answers) =>
    browserPolicyCount(answers, '5.1') >= 2 && browserPolicyCount(answers, '6.1') >= 2,
  evidencePrivacyVisible: browserPolicyEvidencePrivacyVisible,
  notificationEventsRelevant: browserPolicyNotificationEventsRelevant,
  unsupportedCapabilityRelevant: browserPolicyUnsupportedCapabilityRelevant,
  storedBrowserDataExists: browserPolicyStoredBrowserDataExists,
};

export function browserPolicyForestSourceSettingIds(): ReadonlyMap<BrowserPolicyQuestionId, readonly string[]> {
  const settingsByQuestion = new Map<BrowserPolicyQuestionId, string[]>();
  BrowserPolicyQuestionIds.forEach((id) => settingsByQuestion.set(id, []));
  browserControlFullCatalogSettings().forEach((setting) => {
    settingsByQuestion
      .get(browserPolicyQuestionIdForSetting(setting.uiTab, setting.sourceText))
      ?.push(setting.settingId);
  });
  return new Map([...settingsByQuestion.entries()].map(([id, settingIds]) => [id, settingIds.sort()]));
}

function conditionsMatch(conditions: readonly BrowserPolicyCondition[], answers: BrowserPolicyAnswerMap): boolean {
  return conditions.every((condition) => browserPolicyConditionMatches(condition, answers));
}

function conditionListMatches(
  conditions: readonly BrowserPolicyCondition[] | undefined,
  answers: BrowserPolicyAnswerMap
): boolean {
  return conditions !== undefined && conditions.length > 0 && conditionsMatch(conditions, answers);
}

function browserPolicyQuestionIdForSetting(uiTab: string, sourceText: string): BrowserPolicyQuestionId {
  const text = sourceText.toLowerCase();
  if (uiTab === 'ai' || browserPolicySourceTextIncludesAny(text, BrowserPolicyAiSourceTerms)) {
    return 'A1';
  }
  if (uiTab === 'schedule') {
    return browserPolicyScheduleQuestionIdForSetting(text);
  }
  return (
    BrowserPolicyQuestionIdByUiTab.get(uiTab) ??
    browserPolicyQuestionIdForSourceText(text) ??
    BrowserPolicyFallbackQuestionId
  );
}

const BrowserPolicyAiSourceTerms = [' ai ', 'classification'] as const;
const BrowserPolicyScheduleBudgetSourceTerms = ['budget', 'quota'] as const;
const BrowserPolicyFallbackQuestionId = '1.2' satisfies BrowserPolicyQuestionId;
const BrowserPolicyQuestionIdByUiTab = new Map<string, BrowserPolicyQuestionId>([
  ['audit', '18.2'],
  ['data', '17.1'],
  ['reports', '14.1'],
  ['approvals', '12.1'],
  ['setup', '16.1'],
  ['platform', '15.2'],
]);
const BrowserPolicyQuestionSourceRules: readonly {
  readonly terms: readonly string[];
  readonly questionId: BrowserPolicyQuestionId;
}[] = [
  { terms: ['download'], questionId: '9.1' },
  { terms: ['search'], questionId: '7.1' },
  { terms: ['video', 'channel'], questionId: '8.1' },
  { terms: ['managed browser', 'profile', 'extension'], questionId: '3.1' },
  { terms: ['unmanaged', 'bypass', 'tor', 'portable'], questionId: '4.1' },
  { terms: ['url', 'domain', 'category', 'rule'], questionId: '5.1' },
  { terms: ['evidence', 'proof', 'privacy'], questionId: '13.1' },
  { terms: ['browser', 'discover', 'coverage'], questionId: '2.1' },
];

function browserPolicyScheduleQuestionIdForSetting(text: string): BrowserPolicyQuestionId {
  return browserPolicySourceTextIncludesAny(text, BrowserPolicyScheduleBudgetSourceTerms) ? '11.2' : '10.1';
}

function browserPolicyQuestionIdForSourceText(text: string): BrowserPolicyQuestionId | undefined {
  return BrowserPolicyQuestionSourceRules.find((rule) => browserPolicySourceTextIncludesAny(text, rule.terms))
    ?.questionId;
}

function browserPolicySourceTextIncludesAny(text: string, terms: readonly string[]): boolean {
  return terms.some((term) => text.includes(term));
}

File: crates/browser-core/src/browser_policy_questionnaire_forest.rs
use std::collections::{BTreeMap, HashMap};
#[path = "browser_policy_questionnaire_forest_answers.rs"]
mod browser_policy_questionnaire_forest_answers;
#[path = "browser_policy_questionnaire_forest_logic.rs"]
mod browser_policy_questionnaire_forest_logic;
#[path = "browser_policy_questionnaire_forest_question_ids.rs"]
mod browser_policy_questionnaire_forest_question_ids;

use self::browser_policy_questionnaire_forest_answers::*;
use self::browser_policy_questionnaire_forest_logic::*;
use self::browser_policy_questionnaire_forest_question_ids::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum BrowserPolicyCondition {
    AnswerEquals {
        question_id: String,
        option_id: String,
    },
    AnswerIncludes {
        question_id: String,
        option_id: String,
    },
    AnswerIncludesAny {
        question_id: String,
        option_ids: Vec<String>,
    },
    AnswerHasAnySelected {
        question_id: String,
    },
    ComputedFlag {
        flag_id: String,
    },
    All {
        conditions: Vec<BrowserPolicyCondition>,
    },
    Any {
        conditions: Vec<BrowserPolicyCondition>,
    },
    Not {
        condition: Box<BrowserPolicyCondition>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BrowserPolicyQuestionTemplate {
    pub id: String,
    pub surface: String,
    pub show_when: Vec<BrowserPolicyCondition>,
    pub never_show_when: Vec<BrowserPolicyCondition>,
    pub disabled_when: Vec<BrowserPolicyCondition>,
    pub readonly_when: Vec<BrowserPolicyCondition>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BrowserPolicyQuestionStateTemplate {
    pub visible: bool,
    pub disabled: bool,
    pub readonly: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BrowserPolicyCatalogSettingTemplate {
    pub setting_id: String,
    pub ui_tab: String,
    pub source_text: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct BrowserPolicyAnswerMap(BTreeMap<String, Vec<String>>);

impl BrowserPolicyAnswerMap {
    pub fn new(value: BTreeMap<String, Vec<String>>) -> Self {
        Self(value)
    }

    pub fn get(&self, key: &str) -> Option<&Vec<String>> {
        self.0.get(key)
    }

    pub fn into_inner(self) -> BTreeMap<String, Vec<String>> {
        self.0
    }
}

impl From<BTreeMap<String, Vec<String>>> for BrowserPolicyAnswerMap {
    fn from(value: BTreeMap<String, Vec<String>>) -> Self {
        Self::new(value)
    }
}

impl From<BrowserPolicyAnswerMap> for BTreeMap<String, Vec<String>> {
    fn from(value: BrowserPolicyAnswerMap) -> Self {
        value.0
    }
}

pub fn browser_policy_question_state(
    question: &BrowserPolicyQuestionTemplate,
    answers: &BrowserPolicyAnswerMap,
) -> BrowserPolicyQuestionStateTemplate {
    let never_show = condition_list_matches(&question.never_show_when, answers);
    let show = question.show_when.is_empty() || conditions_match(&question.show_when, answers);
    let visible = show && !never_show;

    BrowserPolicyQuestionStateTemplate {
        visible,
        disabled: visible && condition_list_matches(&question.disabled_when, answers),
        readonly: visible && condition_list_matches(&question.readonly_when, answers),
    }
}

pub fn browser_policy_visible_question_ids(
    questions: &[BrowserPolicyQuestionTemplate],
    question_ids: &[String],
    compact_order: &[String],
    answers: &BrowserPolicyAnswerMap,
    surface: &str,
) -> Vec<String> {
    let question_by_id: HashMap<&str, &BrowserPolicyQuestionTemplate> = questions
        .iter()
        .map(|question| (question.id.as_str(), question))
        .collect();
    let ordered_ids: Vec<&String> = if surface == "ai" {
        question_ids
            .iter()
            .filter(|question_id| question_id.starts_with('A'))
            .collect()
    } else {
        compact_order.iter().collect()
    };

    ordered_ids
        .into_iter()
        .filter_map(|question_id| question_by_id.get(question_id.as_str()).copied())
        .filter(|question| question.surface == surface)
        .filter(|question| browser_policy_question_state(question, answers).visible)
        .map(|question| question.id.clone())
        .collect()
}

pub fn browser_policy_forest_source_setting_ids(
    question_ids: &[String],
    settings: &[BrowserPolicyCatalogSettingTemplate],
) -> BTreeMap<String, Vec<String>> {
    let mut settings_by_question: BTreeMap<String, Vec<String>> = question_ids
        .iter()
        .map(|question_id| (question_id.clone(), Vec::new()))
        .collect();

    for setting in settings {
        let question_id =
            browser_policy_question_id_for_setting(&setting.ui_tab, &setting.source_text);
        if let Some(setting_ids) = settings_by_question.get_mut(&question_id) {
            setting_ids.push(setting.setting_id.clone());
        }
    }

    for setting_ids in settings_by_question.values_mut() {
        setting_ids.sort();
    }

    settings_by_question
}

pub fn browser_policy_questionnaire_forest_typescript() -> &'static str {
    include_str!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../browser-core-generated/browser_policy_questionnaire_forest.ts"
    ))
}

pub fn browser_policy_questionnaire_forest_contract_typescript() -> &'static str {
    include_str!("browser_policy_questionnaire_forest_contract.template.txt")
}
